core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1680,45 +1680,61 @@ class KafkaController(val config: KafkaConfig,
-    val partitionsToBeAdded = partitionReplicaAssignment.filter { case (topicPartition, _) =>
+    val newPartitionsToBeAdded = partitionReplicaAssignment.filter { case (topicPartition, _) =>
       controllerContext.partitionReplicaAssignment(topicPartition).isEmpty
     }
 
+    val existingPartitionsInContext = partitionReplicaAssignment.dropWhile(partitionAndReplica => newPartitionsToBeAdded.contains(partitionAndReplica._1))
+    val oldPartitionsToBeModified = existingPartitionsInContext.filter{ case (topicPartition, _) =>
+      controllerContext.partitionReplicaAssignment(topicPartition).diff(partitionReplicaAssignment(topicPartition).replicas).nonEmpty
+    }
+
     if (topicDeletionManager.isTopicQueuedUpForDeletion(topic)) {
-      if (partitionsToBeAdded.nonEmpty) {
+      if (newPartitionsToBeAdded.nonEmpty) {
         warn("Skipping adding partitions %s for topic %s since it is currently being deleted"
-          .format(partitionsToBeAdded.map(_._1.partition).mkString(","), topic))
+          .format(newPartitionsToBeAdded.map(_._1.partition).mkString(","), topic))
+
+        val existingPartitionsInZk = zkClient.getChildren(TopicPartitionsZNode.path(topic))
+        val existingPartitionReplicaAssignment = partitionReplicaAssignment
+          .filter(p => existingPartitionsInZk.contains(p._1.partition.toString))
+          .map { case (tp, _) =>
+            tp -> controllerContext.partitionFullReplicaAssignment(tp)
+          }.toMap
 
-        restorePartitionReplicaAssignment(topic, partitionReplicaAssignment)
+        restorePartitionReplicaAssignment(topic, existingPartitionReplicaAssignment)
       } else {
         // This can happen if existing partition replica assignment are restored to prevent increasing partition count during topic deletion
         info("Ignoring partition change during topic deletion as no new partitions are added")
       }
-    } else if (partitionsToBeAdded.nonEmpty) {
-      info(s"New partitions to be added $partitionsToBeAdded")
-      partitionsToBeAdded.forKeyValue { (topicPartition, assignedReplicas) =>
+    } else if (oldPartitionsToBeModified.nonEmpty) {
+      warn("Skipping modifying existing partitions %s for topic %s, will restore their replica assignment by cache in controllerContext"
+        .format(oldPartitionsToBeModified.map(_._1.partition).mkString(","), topic))
+      val restoreOldPartitionReplicaAssignment = partitionReplicaAssignment.map { case (topicPartition: TopicPartition, assignedReplicas: ReplicaAssignment) =>
+        if (oldPartitionsToBeModified.contains(topicPartition)) {
+          (topicPartition, ReplicaAssignment(controllerContext.partitionReplicaAssignment(topicPartition), Seq.empty, Seq.empty))
+        } else {
+          (topicPartition, assignedReplicas)
+        }
+      }
+      restorePartitionReplicaAssignment(topic, restoreOldPartitionReplicaAssignment)
+    } else if (newPartitionsToBeAdded.nonEmpty) {
```

Comment:
It's possible that both oldPartitionsToBeModified and newPartitionsToBeAdded are nonEmpty and we need to handle that case.

Reply:
> It's possible that both oldPartitionsToBeModified and newPartitionsToBeAdded are nonEmpty and we need to handle that case.

when existing partitions assignments are modified, it will restore existing partitions assignments to zk, if new partitions are added, new partitions assignments will be restored in zk together with the existing partition to trigger the partition modification event again to complete the addition of the new partitions.